arch: arm: cortex_m: Add CPU DWT feature symbol

stephanosio opened the pull request:
```
This commit adds a Kconfig symbol for specifying whether the SoC
implements the CPU DWT feature.

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

| SoC                         | DWT  | Note / Reference                                                  |
|-----------------------------|------|-------------------------------------------------------------------|
| arm/beetle                  | Yes | Core-sight SSE-100 specifies DWT is supported                                |
| arm/mps2                    | No   | Accessing DWT on QEMU MPS2 generates a bus fault                  |
| arm/musca_a                 | Yes | @karl-zh DWT is supported                                |
| arm/musca_b1                | Yes | @karl-zh DWT is supported                                |
| atmel_sam/sam3x             | Yes  | Datasheet specifies that DWT is supported                         |
| atmel_sam/sam4s             | Yes  | Datasheet specifies that DWT is supported                         |
| atmel_sam/same70            | Yes  | Datasheet specifies that DWT is supported                         |
| atmel_sam0/samd20           | No   | ARMv6-M devices do not support DWT                                |
| atmel_sam0/samd21           | No   | ARMv6-M devices do not support DWT                                |
| atmel_sam0/samd51           | Yes  | Datasheet specifies that DWT is supported                         |
| atmel_sam0/same51           | Yes  | Datasheet specifies that DWT is supported                         |
| atmel_sam0/same53           | Yes  | Datasheet specifies that DWT is supported                         |
| atmel_sam0/same54           | Yes  | Datasheet specifies that DWT is supported                         |
| atmel_sam0/samr21           | No   | ARMv6-M devices do not support DWT                                |
| cypress/psoc6               | Yes  | Datasheet specifies that DWT is supported                         |
| microchip_mec/mec1501       | @ioannisg   | Datasheet specifies that DWT is supported  |
| microchip_mec/mec1701       | Yes  | Datasheet specifies that DWT is supported                         |
| nordic_nrf/nrf51            | No   | ARMv6-M devices do not support DWT                                |
| nordic_nrf/nrf52/nrf5281x   | No   | Datasheet specifies that DWT is not supported                     |
| nordic_nrf/nrf52/*          | Yes  | Datasheet specifies that DWT is supported                         |
| nordic_nrf/nrf53            | Yes  | Datasheet specifies that DWT is supported                         |
| nordic_nrf/nrf91            | Yes  | Datasheet specifies that DWT is supported                         |
| nxp_imx/mcimx6x_m4          | Yes @agansari | Datasheet specifies that DWT is supported                                |
| nxp_imx/mcimx7_m4           | Yes  | Reference Manual specifies that DWT is supported                  |
| nxp_imx/rt                  | Yes  | Reference Manual specifies that DWT is supported                  |
| nxp_kinetis/k2x             | Yes  | Reference Manual specifies that DWT is supported                  |
| nxp_kinetis/k6x             | Yes  | Reference Manual specifies that DWT is supported                  |
| nxp_kinetis/k8x             | Yes  | Reference Manual specifies that DWT is supported                  |
| nxp_kinetis/ke1xf           | Yes  | Datasheet specifies that DWT is supported                         |
| nxp_kinetis/kl2x            | No   | ARMv6-M devices do not support DWT                                |
| nxp_kinetis/kv5x            | Yes  | Datasheet specifies that DWT is supported                         |
| nxp_kinetis/kwx/(m4)        | Yes  | Datasheet specifies that DWT is supported                         |
| nxp_kinetis/kwx/(m4+)       | No   | ARMv6-M devices do not support DWT                                |
| nxp_lpc/lpc54xxx            | Yes | "LPCOpen Platform for LPC5411x" contains DWT register definitions |
| nxp_lpc/lpc55xxx            | Yes | User Manual specifies that "four data watch-points" are supported |
| silabs_exx32/efm32hg        | No   | ARMv6-M devices do not support DWT                                |
| silabs_exx32/efm32pg12b     | Yes  | Datasheet specifies that DWT is supported                         |
| silabs_exx32/efm32wg        | Yes  | Datasheet specifies that DWT is supported                         |
| silabs_exx32/efm32fg1p      | Yes  | Datasheet specifies that DWT is supported                         |
| silabs_exx32/efm32mg12p     | Yes  | Datasheet specifies that DWT is supported                         |
| st_stm32/stm32f0            | No   | ARMv6-M devices do not support DWT                                |
| st_stm32/stm32f1            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32f2            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32f3            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32f4            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32f7            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32g0            | No   | ARMv6-M devices do not support DWT                                |
| st_stm32/stm32g4            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32h7            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32l0            | No   | ARMv6-M devices do not support DWT                                |
| st_stm32/stm32l1            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32l4            | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32mp1           | Yes  | Reference Manual specifies that DWT is supported                  |
| st_stm32/stm32wb            | Yes  | Reference Manual specifies that DWT is supported                  |
| ti_lm3s6965                 | Yes  | Datasheet specifies that DWT is supported                         |
| ti_simplelink/cc13x2_cc26x2 | Yes  | Technical Reference Manual specifies that DWT is supported        |
| ti_simplelink/cc32xx        | Yes  | Technical Reference Manual specifies that DWT is supported        |
| ti_simplelink/cc2650        | Yes  | Technical Reference Manual specifies that DWT is supported        |
| ti_simplelink/mps432p4xx    | Yes  | Technical Reference Manual specifies that DWT is supported        |

Supersedes #21636

stephanosio committed:
arch: arm: cortex_m: Add CPU DWT feature symbol

This commit adds a Kconfig symbol for specifying whether the SoC
implements the CPU DWT feature.

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: atmel_sam: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: psoc6: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: microchip_mec: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: nordic_nrf: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: nxp_imx: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: nxp_kinetis: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: nxp_lpc: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: silabs_exx32: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: st_stm32: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: ti_lm3s6965: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: ti_simplelink: Select CPU DWT feature symbol

The Data Watchpoint and Trace (DWT) is an optional debug unit for the
Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+) that provides
watchpoints, data tracing and system profiling capabilities.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

ioannisg reviewed (approved):
This looks fine to me.
Confirmed: for nRF family and atsam3x,4x

erwango reviewed (changes requested):
From the register manuals I've checked, DWT is actually supported on all STM32 devices (even M0/M0+).
So support should be added on F0/G0/L0 series.

stephanosio commented:
> From the register manuals I've checked, DWT is actually supported on all STM32 devices (even M0/M0+).
> So support should be added on F0/G0/L0 series.

While ARMv6-M (Cortex-M0/-M0+) architecture does define a peripheral named "DWT," this is quite different from the DWT found on the ARMv7-M (Cortex-M3/-M4/-M7) in terms of both feature set and register mapping.

In short, the ARMv6-M "DWT" is a very limited subset of the ARMv7-M DWT and they are not compatible with each other as far as I can tell.

We could technically define DWT feature-set symbols for specifying what features each SoC's "DWT" has, but I find that to be somewhat overdoing it as the DWT is mainly used from the application code for its cycle counter register (CYCCNT), which can be useful for benchmarking. Note that this register is not available in the ARMv6-M "DWT."

erwango commented:
@stephanosio, ok thanks for the clarification. So it might be worth to clarify this in symbol help section so people don't mix up. In soc refence manuals, the distinction is not available so we can be easily conuised.

stephanosio commented:
> @stephanosio, ok thanks for the clarification. So it might be worth to clarify this in symbol help section so people don't mix up. In soc reference manuals, the distinction is not available so we can be easily confused.

@erwango I have added comments regarding this in the symbol help.

agansari commented on soc/arm/nxp_lpc/lpc54xxx/Kconfig.soc:
```diff
@@ -10,6 +10,7 @@ choice
 config SOC_LPC54114_M4
 	bool "SOC_LPC54114_M4"
 	select CPU_CORTEX_M4
+	select CPU_CORTEX_M_HAS_DWT
```
Add this for LPC55xxx as well, I can confirm from DWT is available.
Reference manual ch.52 (Debug subsystem) lists "data watch-points" and SDK examples.

stephanosio replied:
Done

karl-zh commented:
Hi,
Depending on the description from C3.2 of Arm® Cortex®‑M33 Processor TRM r0p4. Musca A and Musca B1 both have DWT included.

ioannisg commented:
Beetle is based on SSE-100.
Coresight SSE-100 specifies that SWT is supported:
https://static.docs.arm.com/ddi0551/c/corelink_sse100_subsystem_r0p1_technical_reference_manual_DDI0551C.pdf?_ga=2.114675527.1906771350.1579508973-682412982.1572364770

ioannisg committed:
soc: arm: arm: Select CPU DWT feature symbol for ARM Cortex-M SoCs

The Data Watchpoint and Trace (DWT) is an optional debug unit
for the Cortex-M family cores (except ARMv6-M; i.e. M0 and M0+)
that provides watchpoints, data tracing and system profiling
capabilities. We select it for the ARM Cortex-M SoCs
- Beetle
- Musca A, Musca B
since it is present in these SoCs.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@scottwcpg there's some info missing for 1501, I wonder if you could take a quick look.

agansari commented on soc/arm/nxp_imx/mcimx7_m4/Kconfig.series:
```diff
@@ -6,6 +6,7 @@
 config SOC_SERIES_IMX7_M4
 	bool "i.MX7 M4 Core Series"
 	select CPU_CORTEX_M4
+	select CPU_CORTEX_M_HAS_DWT
```
Add this to mcimx6x_m4 as well.
Reference: IMX6SXRM Chapter 13.7 ARM Cortex M4 Platform/Debug Configuration

ioannisg replied:
Done @agansari , thanks for taking a look

ioannisg commented:
> @scottwcpg there's some info missing for mec1501, I wonder if you could take a quick look.

Edit: Actually, just checked, DWT is present on MEC150X according to the data sheet: 
https://github.com/MicrochipTech/CPGZephyrDocs/blob/master/MEC1501/MEC1501_Datasheet.pdf
